Improve error message for unknown fields of inferred types

# Problem

Trying the code:
```rs
fn main() {
    let _f = |x| x.foo;
}
```
Fails with the error
```
error: Type _ has no member named x
  ┌─ /.../src/main.nr:3:20
  │
3 │     let _f = |foo| foo.x;
  │                    ---

error: aborting due to 1 previous errors
```
which can be made clearer to users that a type annotation should be added before that point.

# Solution

We should special-case this error to suggest adding a type annotation if a user tries to do field access on a type variable.

For reference, rustc's error message here is:
```rs
error[E0282]: type annotations needed
 --> src/main.nr:3:15
  |
3 |     let _f = |foo| foo.x;
  |               ^^^  --- type must be known at this point
  |
help: consider giving this closure parameter an explicit type
  |
3 |     let _f = |foo: _| foo.x;
  |                  +++
```

--- a/crates/noirc_frontend/src/hir/type_check/errors.rs
+++ b/crates/noirc_frontend/src/hir/type_check/errors.rs
@@ -99,8 +99,8 @@ impl From<TypeCheckError> for Diagnostic {
                 span,
             ),
             TypeCheckError::TypeAnnotationsNeeded { span } => Diagnostic::simple_error(
-                "Type annotations needed before this point.".to_string(),
-                "Try adding a type annotation for this value before it is used here".to_string(),
+                "Expression type is ambiguous".to_string(),
+                "Type must be known at this point".to_string(),
                 span,
             ),
         }

--- a/crates/noirc_frontend/src/hir/type_check/errors.rs
+++ b/crates/noirc_frontend/src/hir/type_check/errors.rs
@@ -32,6 +32,8 @@ pub enum TypeCheckError {
         second_type: String,
         second_index: usize,
     },
+    #[error("Cannot infer type of expression, type annotations needed before this point")]
+    TypeAnnotationsNeeded { span: Span },
 }
 
 impl TypeCheckError {
@@ -94,6 +96,11 @@ impl From<TypeCheckError> for Diagnostic {
             TypeCheckError::PublicReturnType { typ, span } => Diagnostic::simple_error(
                 "Functions cannot declare a public return type".to_string(),
                 format!("return type is {typ}"),
+                span,
+            ),
+            TypeCheckError::TypeAnnotationsNeeded { span } => Diagnostic::simple_error(
+                "Type annotations needed before this point.".to_string(),
+                "Try adding a type annotation for this value before it is used here".to_string(),
                 span,
             ),
         }

--- a/crates/noirc_frontend/src/hir/type_check/expr.rs
+++ b/crates/noirc_frontend/src/hir/type_check/expr.rs
@@ -707,9 +707,8 @@ pub fn check_member_access(
     // If we get here the type has no field named 'access.rhs'.
     // Now we specialize the error message based on whether we know the object type in question yet.
     if let Type::TypeVariable(..) = &lhs_type {
-        errors.push(TypeCheckError::TypeAnnotationsNeeded {
-            span: interner.expr_span(&access.lhs),
-        });
+        errors
+            .push(TypeCheckError::TypeAnnotationsNeeded { span: interner.expr_span(&access.lhs) });
     } else if lhs_type != Type::Error {
         errors.push(TypeCheckError::Unstructured {
             msg: format!("Type {lhs_type} has no member named {}", access.rhs),

--- a/crates/noirc_frontend/src/hir/type_check/expr.rs
+++ b/crates/noirc_frontend/src/hir/type_check/expr.rs
@@ -689,34 +689,32 @@ pub fn check_member_access(
 ) -> Type {
     let lhs_type = type_check_expression(interner, &access.lhs, errors).follow_bindings();
 
-    match &lhs_type {
-        Type::Struct(s, args) => {
-            let s = s.borrow();
-            if let Some((field, index)) = s.get_field(&access.rhs.0.contents, args) {
+    if let Type::Struct(s, args) = &lhs_type {
+        let s = s.borrow();
+        if let Some((field, index)) = s.get_field(&access.rhs.0.contents, args) {
+            interner.set_field_index(expr_id, index);
+            return field;
+        }
+    } else if let Type::Tuple(elements) = &lhs_type {
+        if let Ok(index) = access.rhs.0.contents.parse::<usize>() {
+            if index < elements.len() {
                 interner.set_field_index(expr_id, index);
-                return field;
+                return elements[index].clone();
             }
         }
-        Type::Tuple(elements) => {
-            if let Ok(index) = access.rhs.0.contents.parse::<usize>() {
-                if index < elements.len() {
-                    interner.set_field_index(expr_id, index);
-                    return elements[index].clone();
-                }
-            }
-        }
-        Type::TypeVariable(..) => {
-            errors.push(TypeCheckError::TypeAnnotationsNeeded {
-                span: interner.expr_span(&access.lhs),
-            });
-        }
-        Type::Error => (),
-        other => {
-            errors.push(TypeCheckError::Unstructured {
-                msg: format!("Type {other} has no member named {}", access.rhs),
-                span: interner.expr_span(&access.lhs),
-            });
-        }
+    }
+
+    // If we get here the type has no field named 'access.rhs'.
+    // Now we specialize the error message based on whether we know the object type in question yet.
+    if let Type::TypeVariable(..) = &lhs_type {
+        errors.push(TypeCheckError::TypeAnnotationsNeeded {
+            span: interner.expr_span(&access.lhs),
+        });
+    } else if lhs_type != Type::Error {
+        errors.push(TypeCheckError::Unstructured {
+            msg: format!("Type {lhs_type} has no member named {}", access.rhs),
+            span: interner.expr_span(&access.lhs),
+        });
     }
 
     Type::Error

--- a/crates/noirc_frontend/src/hir/type_check/expr.rs
+++ b/crates/noirc_frontend/src/hir/type_check/expr.rs
@@ -687,28 +687,36 @@ pub fn check_member_access(
     expr_id: ExprId,
     errors: &mut Vec<TypeCheckError>,
 ) -> Type {
-    let lhs_type = type_check_expression(interner, &access.lhs, errors);
+    let lhs_type = type_check_expression(interner, &access.lhs, errors).follow_bindings();
 
-    if let Type::Struct(s, args) = &lhs_type {
-        let s = s.borrow();
-        if let Some((field, index)) = s.get_field(&access.rhs.0.contents, args) {
-            interner.set_field_index(expr_id, index);
-            return field;
-        }
-    } else if let Type::Tuple(elements) = &lhs_type {
-        if let Ok(index) = access.rhs.0.contents.parse::<usize>() {
-            if index < elements.len() {
+    match &lhs_type {
+        Type::Struct(s, args) => {
+            let s = s.borrow();
+            if let Some((field, index)) = s.get_field(&access.rhs.0.contents, args) {
                 interner.set_field_index(expr_id, index);
-                return elements[index].clone();
+                return field;
             }
         }
-    }
-
-    if lhs_type != Type::Error {
-        errors.push(TypeCheckError::Unstructured {
-            msg: format!("Type {lhs_type} has no member named {}", access.rhs),
-            span: interner.expr_span(&access.lhs),
-        });
+        Type::Tuple(elements) => {
+            if let Ok(index) = access.rhs.0.contents.parse::<usize>() {
+                if index < elements.len() {
+                    interner.set_field_index(expr_id, index);
+                    return elements[index].clone();
+                }
+            }
+        }
+        Type::TypeVariable(..) => {
+            errors.push(TypeCheckError::TypeAnnotationsNeeded {
+                span: interner.expr_span(&access.lhs),
+            });
+        }
+        Type::Error => (),
+        other => {
+            errors.push(TypeCheckError::Unstructured {
+                msg: format!("Type {other} has no member named {}", access.rhs),
+                span: interner.expr_span(&access.lhs),
+            });
+        }
     }
 
     Type::Error

